jzaia_files/main.cpp

```diff
@@ -0,0 +1,291 @@
+}
+
+/**
+ * @brief The third testing oracle for Grover's
+ *
+ * A 17-qubit test oracle for Grover's algorithm. Applies a pauli-X to the
+ * rightmost qubit if the leftmost 16 qubits are in the state |0011001100110011>
+ *
+ * @param sv The statevector to apply the oracle to. Must be 17 qubits
+ */
+void oracle3(StateVectorLQubitManaged<double> &sv) {
+    // Sanity check statevector
+    assert(sv.getNumQubits() == ORACLE3_QUBITS);
+
+    // Define controls to be used for applying the X gate
+    std::vector<size_t> controls(ORACLE3_QUBITS-1);
+    std::iota(controls.begin(), controls.end(), 0);
+    std::vector<bool> control_vals = ORACLE3_EXPECTED;
+
+    // Apply the X gate to the ancilla, controlled on the chosen bitstring
+    GateImplementationsLM::applyNCPauliX(sv.getData(),
+                                         sv.getNumQubits(),
+                                         controls,
+                                         control_vals,
+                                         {ORACLE3_QUBITS-1},
+                                         false);
+}
+
+
+/**
+ * @brief Overall function for running Grover's algorithm on a chosen oracle
+ *
+ * Run Grover's algorithm from start to finish. Prepares a statevector, and
+ * repeats state selection and amplitude-amplification for sqrt(N) iterations
+ * (where N = 2^(# of non-ancilla qubits)). This implementation assumes that
+ * the oracle always picks precisly 1 state (rather than an arbitrary number).
+ *
+ * @param oracle A black-box function that acts on a created statevector
+ * @param num_qubits The number of qubits in the circuit the oracle acts
+ *        on (includes the ancilla)
+ */
+void run_experiment(void (*oracle) (StateVectorLQubitManaged<double> &),
+                    const size_t num_qubits) {
+    // Create the statevector for this circuit
+    // NOTE: qubits are numbered left-to-right order
+    StateVectorLQubitManaged<double> sv(num_qubits);
+
+    // Set up the circuit by placing into uniform superposition
+    groversSetup(sv);
+
+    // Apply (approx) sqrt(N) repetitions of state selection and amp-amp
+    const size_t nreps = static_cast<size_t>(sqrt(1llu << (num_qubits-1)));
+    for (size_t reps = nreps; reps > 0; --reps) {
+        // Apply the oracle to apply a phase of -1 to desired state
+        oracle(sv);
+        // Perform amp-amp by reflecting over |+++...+>
+        groversMirror(sv);
+    }
+
+    // Set up measurements
+    Measurements<StateVectorLQubitManaged<double>> Measurer(sv);
+    // Vector to store the most common measurement outcome
+    std::vector<bool> common_result(num_qubits - 1, false);
+
+    // Perform a "measurement" by taking the expected value of this qubit over multiple runs
+    for (size_t obs_wire=0; obs_wire < num_qubits - 1; ++obs_wire) {
+        NamedObs<StateVectorLQubitManaged<double>> obs("PauliZ", {obs_wire});
+        double result = Measurer.expval(obs);
+        common_result[obs_wire] = (result < 0);
+    }
```

Comment:
Can this be rewritten in a functional (STL) way?

Reply:
Same disclaimer as the prior comment: To be honest, I'm not as familiar with C++ as I am with plain C, so I had to look into some STL options.

But after a quick glance through some fuilt-in functions. It looks like the `transform` function is the one for the job here. I rewrote that snippet (plus a few lines before it) as:
```C++
// Set up measurements
Measurements<StateVectorLQubitManaged<double>> Measurer(sv);

std::vector<size_t> wires(num_qubits - 1);
std::iota(wires.begin(), wires.end(), 0);
// Vector to store the most common measurement outcome
std::vector<bool> common_result(num_qubits - 1, false);

std::transform(wires.begin(), wires.end(), common_result.begin(), [&Measurer](size_t wire){
    NamedObs<StateVectorLQubitManaged<double>> obs("PauliZ", {wire});
    return Measurer.expval(obs) < 0;
});
```

I've written functional programs before, in languages like Rust, SML, and Lisp (even arguably some python with lambdas). I'm not as used to functional programming in C++, but it seems like it would be relatively easy to pick up.